topics/bloomfilters.md

```diff
@@ -0,0 +1,108 @@
+
+## Basic Bloom commands
+
+* `BF.ADD` adds an item to a bloom filter
+* `BF.CARD` returns the cardinality of a bloom filter
+* `BF.EXISTS` checks if an item has been added to a bloom filter
+* `BF.INFO` returns information about a bloom filter
+
+See the [complete list of bloom filter commands](../commands/#bloom).
+
+## Common use cases for bloom filters
+
+### Financial fraud detection
+
+Bloom filters can help answer the question "Has this card been flagged as stolen?", use a bloom filter that has cards reported stolen added to it. Check a card on use that it is not present in the bloom filter. If it isn't then the card is not marked as stolen, if present then a check to the main database can happen or deny the purchase.
+
+### Ad placement
+
+Bloom filters can help answer the following questions to advertisers:
+* Has the user already seen this ad?
+* Has the user already bought this product?
+
+Use a Bloom filter for every user, storing all bought products. The recommendation engine can then suggest a new product and checks if the product is in the user's Bloom filter.
+
+* If no, the ad is shown to the user and is added to the Bloom filter.
+* If yes, the process restarts and repeats until it finds a product that is not present in the filter.
+
+### Check if URL's are malicious
+
+Bloom filters can answer the question "is a URL malicious?". Any URL inputted would be checked against a malicious URL bloom filter. 
+
+* If no then we allow access to the site
+* If yes then we can deny access or perform a full check of the URL
+
+### Check if a username is taken
+
+Bloom filters can answer the question: Has this username/email/domain name/slug already been used?
+
+For example for usernames. Use a Bloom filter for every username that has signed up. A new user types in the desired username. The app checks if the username exists in the Bloom filter.
+
+* If no, the user is created and the username is added to the Bloom filter.
+* If yes, the app can decide to either check the main database or reject the username.
+
+## Default bloom properties
+
+Capacity - 100
+
+Error rate - 0.01
+
+Expansion - 2
+
+As bloom filters have a default expansion of 2 this means all default bloom objects will be scaling. These options are used when not specified explicitly in the commands used to create a new bloom object. For example doing a BF.ADD for a new filter will create a filter with the exact above qualities. These default properties can be configured through configs on the bloom module.
+Example of default bloom objects information:
+
+```
+127.0.0.1:6379> BF.ADD default_filter item
+1
+127.0.0.1:6379> BF.INFO default_filter
+ 1) Capacity
+ 2) (integer) 100
+ 3) Size
+ 4) (integer) 384
+ 5) Number of filters
+ 6) (integer) 1
+ 7) Number of items inserted
+ 8) (integer) 1
+ 9) Error rate
+10) "0.01"
+11) Expansion rate
+12) (integer) 2
+13) Max scaled capacity
+14) (integer) 26214300
+```
+
+## Performance
```

Comment:
It might also be good to talk about the implication of number of scaled out filters vs performance and how it is recommended to choose the capacity and / or expansion rate to better suit their use case in terms of expected number of items to be added